File: frontend/radvis/src/app/shared/models/belag-art.ts
/*
 * Copyright (c) 2023 WPS - Workplace Solutions GmbH
 *
 * Licensed under the EUPL, Version 1.2 or as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

import { EnumOption } from 'src/app/form-elements/models/enum-option';

export enum BelagArt {
  ASPHALT = 'ASPHALT',
  BETON = 'BETON',
  NATURSTEINPFLASTER = 'NATURSTEINPFLASTER',
  BETONSTEINPFLASTER_PLATTENBELAG = 'BETONSTEINPFLASTER_PLATTENBELAG',
  WASSERGEBUNDENE_DECKE = 'WASSERGEBUNDENE_DECKE',
  UNGEBUNDENE_DECKE = 'UNGEBUNDENE_DECKE',
  SONSTIGER_BELAG = 'SONSTIGER_BELAG',
  UNBEKANNT = 'UNBEKANNT',
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace BelagArt {
  export const options: EnumOption[] = Object.keys(BelagArt).map((k: string): EnumOption => {
    switch (k) {
      case BelagArt.ASPHALT:
        return { name: k, displayText: 'Asphalt' };
      case BelagArt.BETON:
        return { name: k, displayText: 'Beton' };
      case BelagArt.NATURSTEINPFLASTER:
        return { name: k, displayText: 'Natursteinpflaster' };
      case BelagArt.BETONSTEINPFLASTER_PLATTENBELAG:
        return { name: k, displayText: 'Betonsteinpflaster oder Plattenbelag' };
      case BelagArt.WASSERGEBUNDENE_DECKE:
        return { name: k, displayText: 'Wassergebundene Decke' };
      case BelagArt.UNGEBUNDENE_DECKE:
        return { name: k, displayText: 'Ungebundene Decke' };
      case BelagArt.SONSTIGER_BELAG:
        return { name: k, displayText: 'Sonstiger Belag' };
      case BelagArt.UNBEKANNT:
        return { name: k, displayText: 'Unbekannt' };
    }
    throw new Error('Beschreibung für enum BelagArt fehlt: ' + k);
  });
}

File: backend/src/main/java/de/wps/radvis/backend/netz/domain/valueObject/BelagArt.java
/*
 * Copyright (c) 2023 WPS - Workplace Solutions GmbH
 *
 * Licensed under the EUPL, Version 1.2 or as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package de.wps.radvis.backend.netz.domain.valueObject;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.NonNull;

@AllArgsConstructor(access = AccessLevel.PRIVATE)
public enum BelagArt {
	//@formatter:off
	ASPHALT("Asphalt"),
	BETON("Beton"),
	NATURSTEINPFLASTER("Natursteinpflaster"),
	BETONSTEINPFLASTER_PLATTENBELAG("Betonsteinpflaster Plattenbelag"),
	WASSERGEBUNDENE_DECKE("Wassergebundene Decke"),
	UNGEBUNDENE_DECKE("Ungebundene Decke (Kies, Split, Sand, Erde, Gras, Spurbahn)"),
	SONSTIGER_BELAG("Sonstiger Belag"),
	WASSERGEBUNDENE_DECKE_MIT_GRUENSTREIFEN("Wassergebundene Decke mit Grünstreifen mittig"),
	ASPHALT_MIT_GRUENSTREIFEN("Asphalt mit Rasengittersteinen / Grünstreifen mittig"),
	BETON_MIT_GRUENSTREIFEN("Beton mit Rasengittersteinen / Grünstreifen mittig"),
	UNBEKANNT("Unbekannt");
	//@formatter:on

	@NonNull
	private final String displayText;

	@Override
	public String toString() {
		return this.displayText;
	}

	public BelagArt nichtUnbekanntOrElse(BelagArt otherValue) {
		return !this.equals(BelagArt.UNBEKANNT) ? this : otherValue;
	}

	public boolean widerspruchZu(BelagArt other) {
		return !(this == BelagArt.UNBEKANNT || other == BelagArt.UNBEKANNT || this.equals(other));
	}
}